src/app/util/mock/attribute-storage.cpp

```diff
@@ -0,0 +1,187 @@
+    Clusters::Globals::Attributes::ClusterRevision::Id, Clusters::Globals::Attributes::FeatureMap::Id, MockAttributeId(1), MockAttributeId(2), MockAttributeId(3),
+    Clusters::Globals::Attributes::ClusterRevision::Id, Clusters::Globals::Attributes::FeatureMap::Id, MockAttributeId(1),
+    Clusters::Globals::Attributes::ClusterRevision::Id, Clusters::Globals::Attributes::FeatureMap::Id, MockAttributeId(1), MockAttributeId(2), MockAttributeId(3), MockAttributeId(4),
+    Clusters::Globals::Attributes::ClusterRevision::Id, Clusters::Globals::Attributes::FeatureMap::Id,
+    Clusters::Globals::Attributes::ClusterRevision::Id, Clusters::Globals::Attributes::FeatureMap::Id
+    // clang-format on
+};
+
+} // namespace
+
+uint16_t emberAfEndpointCount(void)
+{
+    return ArraySize(endpoints);
+}
+
+uint16_t emberAfIndexFromEndpoint(chip::EndpointId endpoint)
+{
+    for (uint16_t i = 0; i < ArraySize(endpoints); i++)
+    {
+        if (endpoints[i] == endpoint)
+        {
+            return i;
+        }
+    }
+    return UINT16_MAX;
+}
+
+uint8_t emberAfClusterCount(chip::EndpointId endpoint, bool server)
+{
+    for (uint16_t i = 0; i < ArraySize(endpoints); i++)
+    {
+        if (endpoints[i] == endpoint)
+        {
+            return clusterCount[i];
+        }
+    }
+    return 0;
+}
+
+uint16_t emberAfGetServerAttributeCount(chip::EndpointId endpoint, chip::ClusterId cluster)
+{
+    uint16_t endpointIndex          = emberAfIndexFromEndpoint(endpoint);
+    uint16_t clusterCountOnEndpoint = emberAfClusterCount(endpoint, true);
+    for (uint16_t i = 0; i < clusterCountOnEndpoint; i++)
+    {
+        if (clusters[i + clusterIndex[endpointIndex]] == cluster)
+        {
+            return attributeCount[i + clusterIndex[endpointIndex]];
+        }
+    }
+    return UINT16_MAX;
+}
+
+uint16_t emberAfGetServerAttributeIndexByAttributeId(chip::EndpointId endpoint, chip::ClusterId cluster,
+                                                     chip::AttributeId attributeId)
+{
+    uint16_t endpointIndex          = emberAfIndexFromEndpoint(endpoint);
+    uint16_t clusterCountOnEndpoint = emberAfClusterCount(endpoint, true);
+    for (uint16_t i = 0; i < clusterCountOnEndpoint; i++)
+    {
+        if (clusters[i + clusterIndex[endpointIndex]] == cluster)
+        {
+            uint16_t clusterAttributeOffset = attributeIndex[i + clusterIndex[endpointIndex]];
+            for (uint16_t j = 0; j < emberAfGetServerAttributeCount(endpoint, cluster); j++)
+            {
+                if (attributes[clusterAttributeOffset + j] == attributeId)
+                {
+                    return j;
+                }
+            }
+            break;
+        }
+    }
+    return UINT16_MAX;
+}
+
+chip::EndpointId emberAfEndpointFromIndex(uint16_t index)
+{
+    return index < ArraySize(endpoints) ? endpoints[index] : app::ConcreteAttributePath::kInvalidEndpointId;
```

Comment:
This also does not match the non-mock impl, which assumes that the index passed in is valid.  This could cover up bugs where something tested against this mock passes in invalid indices.

Probably best to `VerifyOrDie(index < ArraySize(endpoints))` and then just return `endpoints[index]`.